tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/types/family/TType.java

```diff
@@ -17,21 +17,40 @@
 
 package org.tensorflow.types.family;
 
+import org.tensorflow.Tensor;
+
 /**
- * Marker interface for all tensor types.
+ * Common interface for all typed tensors.
  *
- * <p>Tensor types are carried as a generic parameter of the {@link org.tensorflow.Tensor Tensor}
- * class bound by the {@code TType} interface. This generic parameter ensure type-compatibility
- * between operands of a computation at compile-time. For example:
+ * <p>Typed tensors wraps a {@link RawTensor} by mapping their native memory to a n-dimensional
+ * data space allowing direct I/O access from the JVM.</p>
  *
- * <pre>{@code
- * Tensor<TFloat32> tensor1 = TFloat32.ofShape(2, 3, 2);
- * Tensor<TFloat32> tensor2 = TFloat32.ofShape(2, 3, 2);
- * Tensor<TInt32> tensor3 = TInt32.ofShape(2, 3, 2);
+ * <p>Subinterfaces of {@code TType} are propagated as a generic parameter to various entities of
+ * TensorFlow to identify the type of the tensor they carry. For example, a
+ * {@link org.tensorflow.Operand Operand<TFloat32>} is an operand which outputs is a 32-bit floating
+ * point tensor. This parameter ensure type-compatibility between operands of a computation at
+ * compile-time. For example:
  *
+ * <pre>{@code
  * Ops tf = Ops.create();
- * tf.math.add(tf.constant(tensor1), tf.constant(tensor2));  // OK
- * tf.math.add(tf.constant(tensor1), tf.constant(tensor3));  // Compilation failure
+ *
+ * Constant<TFloat32> c1 = tf.array(2.0f, 3.0f, 2.0f);
+ * Constant<TFloat32> c2 = tf.array(1.0f, 2.0f, 3.0f);
+ * Constant<TInt32> c3 = tf.array(2, 3, 2);
+ *
+ * tf.math.add(c1, c2);  // OK
+ * tf.math.add(c1, c3);  // Compilation failure
  * }</pre>
  */
-public interface TType {}
+public interface TType extends Tensor {
```

Comment:
I deliberately tried to leave `TType` unparameterized, partially to reduce verbosity since this interface is carried in many endpoints and containers. If we add a parameter to `TType`, signatures with `T extends TType` now become `T extends TType<U>, U`, where you'll need to carry `U` everywhere as well, unless you turn it into a wildcard but then you lose the plus value of having a parameter.

Both `TType` and `RawTensor` instances are a `Tensor`, but only the former is typed so that a user can access its data properly but requiring him to cast it to a concrete type to do it. This is intentional as it leverages the use of primitive access endpoints to retrieve the data instead of the boxed ones. 

For example, if a user has an instance of `TType<Float>` that extends from `NdArray<Float>`, chances are that he would retrieve its data probably directly via the boxed-type methods like `t.getObject()`, which would return costly `Float` objects. 

But by enforcing a cast from `TType` to `TFloat32` instead, he would retrieve the same data but now being able to use `t.getFloat()` from `FloatNdArray`, which returns `float` primitive values, improving performances overall.

Reply:
Convincing! Perhaps incorporate the essence of this explanation as a comment?